lib/SIL/IR/SILFunctionType.cpp

```diff
@@ -235,8 +235,25 @@ IndexSubset *SILFunctionType::getDifferentiabilityResultIndices() {
       resultIndices.push_back(resultAndIndex.index());
   // Check `inout` parameters.
   for (auto inoutParamAndIndex : enumerate(getIndirectMutatingParameters()))
-    if (inoutParamAndIndex.value().getDifferentiability() !=
-        SILParameterDifferentiability::NotDifferentiable)
+    // FIXME(TF-1305): The `getResults().empty()` condition is a hack.
+    //
+    // Currently, an `inout` parameter can either be:
+    // 1. Both a differentiability parameter and a differentiability result.
+    // 2. `@noDerivative`: neither a differentiability parameter nor a
+    //    differentiability result.
+    // However, there is no way to represent an `inout` parameter that:
+    // 3. Is a differentiability result but not a differentiability parameter.
+    // 4. Is a differentiability parameter but not a differentiability result.
+    //    This case does not yet have clear use cases, so supporting it is a
+    //    non-goal.
```

Comment:
@rxwei mentioned a nuanced point, which is that there are no clear use cases for `inout` parameters that are differentiability parameters but not differentiability results (case 4).

That case is indeed not currently expressible, so we can deprioritize support for it: I clarified that in the comment. Supporting that case would involve something like adding a `results:` clause to `@differentiable` declaration attribute:
```swift
// Pseudo-syntax for supporting case 4:
// `inout` parameter that is a differentiability parameter but not a differentiability result.
@differentiable(wrt: inoutParam, results: return)
func foo(_ inoutParam: inout Float) -> Float {
  return inoutParam
}
```

The focus of this PR is supporting `inout` parameters that are differentiability results but that are not differentiability parameters (case 3), which is currently expressible and reasonable to support.

```swift
// Case 3: `inout` parameter that is a differentiability result but not a differentiability parameter.
@differentiable(wrt: x)
func foo(_ x: Float, _ inoutParam: inout Float) {
  inoutParam = x * x
}
```